src/lib/mina_block/body_reference.ml

```diff
@@ -10,19 +10,42 @@ open Core_kernel
 
 [%%versioned
 module Stable = struct
-  module V1 = struct
-    (* type t = State_body_hash.Stable.V1.t *)
-    type t = unit [@@deriving compare, sexp, yojson]
+  module V2 = struct
+    type t = Blake2.Stable.V1.t * Mina_base.Signature.Stable.V1.t
+    [@@deriving sexp, yojson]
 
     let to_latest = Fn.id
   end
 end]
 
 type t = Stable.Latest.t
 
-[%%define_locally
-Stable.V1.(compare, t_of_sexp, sexp_of_t, of_yojson, to_yojson)]
+[%%define_locally Stable.Latest.(t_of_sexp, sexp_of_t, to_yojson, of_yojson)]
 
-let of_body = Fn.const ()
+let message_for_sign ref :
+    (Snark_params.Tick.Field.t, bool) Random_oracle.Legacy.Input.t =
+  let field_elements = [||] in
+  let bitstrings =
+    [| Blake2.to_raw_string ref |> Blake2.string_to_bits |> Array.to_list |]
+  in
+  { field_elements; bitstrings }
 
-(* let of_body = Body.hash *)
+let sign_reference ~private_key =
+  Fn.compose (Signature_lib.Schnorr.Legacy.sign private_key) message_for_sign
+
+let compute_reference b =
+  let sz = Body.Stable.V1.bin_size_t b in
+  let buf = Bin_prot.Common.create_buf sz in
+  ignore (Body.Stable.V1.bin_write_t buf ~pos:0 b : int) ;
+  Tuple2.get2 @@ Mina_net2.Bitswap_block.blocks_of_data ~max_block_size:1024 buf
```

Comment:
TODO: use the same value as in helper

Reply:
NIT: You can just use `snd` instead of `Tuple2.get`. But I do have a slight preference for the expllicit style of `let _, x = ... in x` for clarity.

Reply:
Is this step actually necessary? Any data that we download from Bitswap comes with the guarantee that it matches the hash since it's CID based storage.